helix-view/src/document.rs

```diff
@@ -236,26 +242,107 @@ impl Document {
     }
 
     // TODO: async fn?
-    pub fn load(
+    fn open_impl(
+        path: PathBuf,
+        encoding: Option<&'static encoding_rs::Encoding>,
+    ) -> Result<(Rope, &'static encoding_rs::Encoding), Error> {
+        use std::{fs::File, io::Read};
+
+        if !path.exists() {
+            return Err(anyhow::anyhow!("path doesn't exist"));
+        }
+
+        // Create two 8kB buffers to write the file into and the decoded contents out to.
+        let mut buf = [0u8; BUF_SIZE];
+        let mut buf_out = [0u8; BUF_SIZE];
+        let mut builder = RopeBuilder::new();
+        let mut file = File::open(&path)?;
+        let (encoding, mut decoder, mut slice) = {
+            let read = file.read(&mut buf)?;
+            let last = read != buf.len();
```

Comment:
I sort of understand, but not really. When it comes to file I/O, wouldn't it be a non-issue, or is it in case a file is being streamed or over the net?

Reply:
@kirawi There's no guarantees from the OS about this for file I/O. In practice it will always fill the buffer, but we can't rely on this. `0` is the correct invariant to check:

> If n is 0, then it can indicate one of two scenarios:
>
>    - This reader has reached its “end of file” and will likely no longer be able to produce bytes. Note that this does not mean that the reader will always no longer be able to produce bytes.
>    - The buffer specified was 0 bytes in length.